### PR TITLE
gravel: expose bootstrap progress

### PR DESCRIPTION
**Pull request by jecluis**
This is extremely fragile, as we're essentially parsing cephadm's
free-form output and associating each step with a rule of thumbs
percentage. Should cephadm change their messaging, this can easily
break.

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

### CODE REVIEW

**Review by votdev — APPROVED** [review]
LGTM and a great improvement.

**Comment by votdev** [comment]
Can't we implement the feature to report the deployment stage in cephadm and contribute it upstream?

File: src/gravel/cephadm/cephadm.py [hunk 1]
```diff
@@ -64,24 +66,50 @@ async def call(self, cmd: str) -> Tuple[str, str, int]:
...
+            ("creating mon", 20),
+            ("waiting for mon to start", 25),
+            ("wrote config", 40),
+            ("creating mgr", 45),
+            ("waiting for mgr to start", 50),
+            ("enabling cephadm module", 60),
+            ("setting orchestrator backend to cephadm", 70),
+            ("adding host", 80),
+            ("bootstrap complete", 100)
+        ]
```
Comment (mgfritch) [inline src/gravel/cephadm/cephadm.py:103]: super fragile, seems like this would justify a unit test with various types of output from cephadm ..

**Review by mgfritch — COMMENTED** [review]
> Can't we implement the feature to report the deployment stage in cephadm and contribute it upstream?

I was wondering about the exact same thing ... not a huge fan of parsing arbitrary output.

**Comment by jecluis** [comment]
> > Can't we implement the feature to report the deployment stage in cephadm and contribute it upstream?
> 
> I was wondering about the exact same thing ... not a huge fan of parsing arbitrary output.

@sebastian-philipp Thoughts? Would this be something that would be interesting upstream?

**Comment by sebastian-philipp** [comment]
This needs serious refactoring in upstream. We have IMO to abandon the "single big blob of python executable" in cephadm first. 

1. https://github.com/ceph/ceph/pull/39141 (having reviews will help)
2. https://github.com/ceph/ceph/pull/37495
3. change the url in https://docs.ceph.com/en/latest/cephadm/install/#install-cephadm to download.ceph.com
4. make cephadm a proper module and zip/package/compile it into an executable. 

And then we can start with this. 

Doing this now would make future refactoring even more complicated and cumbersome.

**Comment by jecluis** [comment]
@sebastian-philipp Ok, so... how can we help make that happen? Is there a path? Issues, tasks, something we can devote people and time to?

**Comment by sebastian-philipp** [comment]
> @sebastian-philipp Ok, so... how can we help make that happen? Is there a path? Issues, tasks, something we can devote people and time to?

Reviewing https://github.com/ceph/ceph/pull/39141 would help me a lot!

**Comment by sebastian-philipp** [comment]
next steps would be to upload cephadm to download.ceph.com instead of letting users wget from the source tree.

**Comment by sebastian-philipp** [comment]
and please review https://github.com/ceph/ceph/pull/39619

**Review by jhesketh — APPROVED** [review]
lgtm except for missing tests

File: src/gravel/cephadm/cephadm.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/gravel/cephadm/cephadm.py:103]: https://github.com/ceph/ceph/pull/40103

Comment (jecluis) [reply src/gravel/cephadm/cephadm.py:103]: oof. Glad it didn't change any of those, eh? :)

**Comment by jecluis** [comment]
@votdev rebased and repushed. Needs a green checkmark to merge.

**Review by votdev — APPROVED** [review]

**Pull request merged**